EIPS/eip-6404.md

```diff
@@ -0,0 +1,813 @@
+---
+eip: 6404
+title: SSZ transactions, receipts, and withdrawals
+description: Migration of MPT commitments for transactions, receipts, and withdrawals to SSZ
+author: Etan Kissling (@etan-status), Vitalik Buterin (@vbuterin)
+discussions-to: https://ethereum-magicians.org/t/eip-6404-ssz-transactions-receipts-and-withdrawals/12783
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-01-30
+requires: 155, 658, 1559, 2718, 2930, 4844, 4895
+---
+
+## Abstract
+
+This EIP defines a migration process of existing Merkle-Patricia Trie (MPT) commitments for transactions, receipts, and withdrawals to SSZ.
+
+## Motivation
+
+While the consensus `ExecutionPayload` and the execution block structure map to each other conceptually, they are encoded differently. This EIP aims to align the encoding of their fields, taking advantage of the more modern SSZ format. This brings several advantages:
+
+1. **Reducing complexity:** Merkle-Patricia Tries (MPT) are hard to work with. Replacing them with SSZ leaves only the state trie in the legacy MPT format.
+
+2. **Better for smart contracts:** The SSZ format is optimized for production and verification of merkle proofs. It allows proving specific fields of containers and allows chunked processing, e.g., to support handling transactions that do not fit into calldata.
+
+3. **Better for light clients:** Light clients with access to the consensus `ExecutionPayload` no longer need to obtain the matching execution block header to verify proofs rooted in `transactions_root` or `withdrawals_root`. Note that the `receipts_root` is currently represented as a MPT root across both execution and consensus.
+
+4. **Reducing ambiguity:** The names `transactions_root` and `withdrawals_root` are currently used to refer to different roots. The execution block header refers to a MPT root, the consensus `ExecutionPayloadHeader` refers to a SSZ root.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### SSZ
+
+#### `Optional[T]`
+
+A new [SSZ type](https://github.com/ethereum/consensus-specs/blob/67c2f9ee9eb562f7cc02b2ff90d92c56137944e1/ssz/simple-serialize.md) is introduced to represent `Optional[T]` values.
+
+- If value is `None`, serialize as `[]`, and merkleize as `List[T, 1]` with length `0`.
+- If value is not `None`, serialize as `T`, and merkleize as `List[T, 1]` with length `1`.
+- Serialize `Optional[T]` as variable-size object (create offset-table entry in enclosing containers).
```

Comment:
Should T be allowed to have min_size 0 (for example `Optional[List[uint8, 1]]`) ? If so its deserialization could result in two different values

Reply:
Will follow up about this with @zah, technical discussion should be in discussions-to thread.

Reply:
@dapplion SSZ discussion split off into EIP-6475: https://ethereum-magicians.org/t/eip-6475-ssz-optional/12891

`Optional[List[...]]` is an illegal type. Use an empty list instead to represent that.